Implement package pinning

florelis opened the pull request:
This PR builds on top of #2769. I can close that PR in favor of this one, or we can merge that one first to make the changes smaller and easier to review :)

Changes in this PR (not in #2769):
* Modified the `CompositePackage` to be able to hold more than one available package, each from a different source. The `GetAvailableVersionKeys()` function now can return versions from multiple sources, so the source field of `PackageVersionKey`  is more relevant now. If the same version is available from multiple sources, they are returned in the order the available packages were added, which matches the previous behavior of just adding the first available package found.
* Added a `PinBehavior` enum which is passed as an argument to an `IPackage`'s functions for getting available versions. Only the `CompositePackage` uses it (as it needs both an installed and an available package) and it is used to determine whether we ignore pinned versions when listing the available versions or not. For example, for `list` we don't care about pins, for `upgrade` we do consider them, and for `upgrade --include-pinned` we care about them unless they have Pinning type.
  + Added this new argument all through the codebase. I'm open to making it have a default value, but I wanted to highlight all the places it is used to be sure I made the appropriate choice on each place.
* Updated existing tests to new behaviors. For example, `CompositeSource` tests now need to ensure that the test packages have a reference to their source as it is used to determine if they are pinned.
* There was already some work for pinning done for the `RequiresExplicitUpgrade` manifest field. This change is mostly independent of that one, as in my mind the pinning from `RequiresExplicitUpgrade` is a property of the installed package, whereas user-defined pins are properties of the available packages from each source, so they are handled at different places.

Closes #476
Related to #2611

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2813)

florelis commented:
Here's a little screenshot of this thing in action:

![image](https://user-images.githubusercontent.com/14323496/210416502-307f681f-e9dd-4aeb-85f6-097cd0dea0e5.png)

Trenly commented on src/AppInstallerCLICore/Workflows/UpdateFlow.cpp:
```diff
@@ -190,19 +196,18 @@ namespace AppInstaller::CLI::Workflow
+                // RequireExplicitUpgrade. While this has the downside of not working with
+                // packages installed from another source, it ensures consistency with the
+                // list of available updates (there we don't have the selected installer)
+                // and at most we will update each package like this once.
```
Does this PR also resolve https://github.com/microsoft/winget-cli/issues/2559#issuecomment-1262712788 ?

florelis replied:
No, it does not. If we make any changes for that, I'd rather have them in a separate PR as this one is already huge.

Trenly replied:
I understand :)
Just not sure where the original bug actually came in, so was hoping it might have been cauggt in the mix. Thank you for the clarification!

Trenly commented on src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:
```diff
@@ -1605,4 +1597,39 @@ Please specify one of them using the --source option to proceed.</value>
+    <comment>{Locked="--force"}</comment>
+  </data>
+  <data name="PinExistsUseForceArg" xml:space="preserve">
+    <value>A pin already exists for this package. Use the --force argument to overwrite it.</value>
```
Is there a reason not to use the `{0}` and the package name here? Just curious, since the `PinDoesNotExist` message does use it

florelis replied:
No reason. I'll change that. It's part of what I meant with "Make this pretty. Lots of places could use better or more messages." in the PR description :)

Trenly commented on src/AppInstallerCommonCore/Errors.cpp:
```diff
@@ -208,6 +208,12 @@ namespace AppInstaller
+            case APPINSTALLER_CLI_ERROR_PIN_DOES_NOT_EXIST:
+                return "There is no pin for the package.";
+            case APPINSTALLER_CLI_ERROR_PACKAGE_HAS_BLOCKING_PIN:
+                return "The package has a blocking pin that prevents from updating.";
```
```suggestion
                return "The package has a blocking pin that prevents upgrade.";
```

florelis commented on src/AppInstallerCLICore/Workflows/PinFlow.cpp:
```diff
@@ -0,0 +1,267 @@
+        context.Add<Execution::Data::Pins>(context.Get<Execution::Data::PinningIndex>()->GetAllPins());
+    }
+
+    void SearchPin(Execution::Context& context)
```
I think I'll remove the comments from the other workflow tasks instead. From what I see, we generally keep the comments in the headers. The others have comments only because I copied the definitions from the header and I forgot to remove them.

Trenly replied:
Works for me!

yao-msft commented on src/AppInstallerRepositoryCore/Public/winget/RepositorySearch.h:
```diff
@@ -243,8 +233,16 @@ namespace AppInstaller::Repository
+        {
+            // Sort using only the version and channel.
+            // The order for the sources depends on the context.
+            return Utility::VersionAndChannel({ Version }, { Channel }) < Utility::VersionAndChannel({ other.Version }, { other.Channel });
```
This code is assuming versions across sources use same semantic. I remember seeing packages that use different versions for Store source and other sources. Should we not allow version comparison if it's from different sources?

yao-msft commented on src/AppInstallerRepositoryCore/Microsoft/PinningIndex.cpp:
```diff
@@ -26,9 +26,24 @@ namespace AppInstaller::Repository::Microsoft
+        const auto indexPath = DefaultIndexPath;
+#endif
+
+        AICLI_LOG(CLI, Info, << "Opening pinning index");
```
```suggestion
        AICLI_LOG(Repo, Info, << "Opening pinning index");

```

yao-msft commented:
>                         // TODO: Needs a whole separate change to fix the fact that we don't support multiple available packages and what the different search behaviors mean

Is this still relevant?

---
Refers to: src/AppInstallerRepositoryCore/CompositeSource.cpp:1293 in 395a9f3. [](commit_id = 395a9f39c1337469ceea6e9bba41833f3805cc11, deletion_comment = False)

yao-msft reviewed (commented):
I think we can add some e2e tests later since after this change, the feature is considiered primarily complete.

yao-msft commented:
By the way, I did not see changes related to --force override behavior for pinning, I thought with user passing in --force, PinBehavior will be IgnorePins, is that not needed in the whole flow?

yao-msft commented on src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:
```diff
@@ -708,14 +709,15 @@ namespace AppInstaller::CLI::Workflow
-                auto latestVersion = match.Package->GetLatestAvailableVersion();
-                bool updateAvailable = match.Package->IsUpdateAvailable();
+                auto latestVersion = match.Package->GetLatestAvailableVersion(pinBehavior);
+                bool updateAvailable = match.Package->IsUpdateAvailable(pinBehavior);
```
I think for `winget list`, we are hiding latest available versions from users forever if a pin is set, maybe we should just apply the pin behavior for upgrade related scenarios

denelon replied:
I think we need to have a mechanism to show if a package is pinned in list so a user could see when new versions are available. The pinning might be temporary to avoid a known bug, and the user might still be looking for any new versions.

I believe the same is true for winget upgrade. We might want a separate section for pinned packages so we can still show if new versions are available for the same reason.

florelis replied:
I added a count of user-pinned packages at the end of the update list, and I'm now showing the latest available version on list even if it is pinned.

yao-msft commented on src/AppInstallerCLICore/Workflows/DependencyNodeProcessor.cpp:
```diff
@@ -41,7 +41,7 @@ namespace AppInstaller::CLI::Workflow
         auto packageId = package->GetProperty(PackageProperty::Id);
         m_nodePackageInstalledVersion = package->GetInstalledVersion();
-        m_nodePackageLatestVersion = package->GetLatestAvailableVersion();
+        m_nodePackageLatestVersion = package->GetLatestAvailableVersion(PinBehavior::ConsiderPins);
```
Does --include-pinned and --force apply here too?

yao-msft commented on src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:
```diff
@@ -861,7 +862,7 @@ namespace AppInstaller::CLI::Workflow
     {
         PackageVersionKey key("", m_version, m_channel);
-        auto requestedVersion = context.Get<Execution::Data::Package>()->GetAvailableVersion(key);
+        auto requestedVersion = context.Get<Execution::Data::Package>()->GetAvailableVersion(key, PinBehavior::ConsiderPins);
```
```suggestion
        auto requestedVersion = context.Get<Execution::Data::Package>()->GetAvailableVersion(key, PinBehavior::IgnorePins);

```

This looks to be only used by show command

florelis replied:
The upgrade command also uses this. I added an argument to the workflow task to consider or not the pins so that `show` can ignore them.

yao-msft commented on src/AppInstallerCommonCore/Public/AppInstallerErrors.h:
```diff
@@ -113,6 +113,7 @@
 #define APPINSTALLER_CLI_ERROR_PIN_ALREADY_EXISTS               ((HRESULT)0x8A150062)
 #define APPINSTALLER_CLI_ERROR_PIN_DOES_NOT_EXIST               ((HRESULT)0x8A150063)
 #define APPINSTALLER_CLI_ERROR_CANNOT_OPEN_PINNING_INDEX        ((HRESULT)0x8A150064)
+#define APPINSTALLER_CLI_ERROR_PACKAGE_HAS_BLOCKING_PIN         ((HRESULT)0x8A150065)
```
This error code is not used.

I think this is a general issue with the current implementation of hiding available versions by pin behavior. User will not be notified why a package upgrade is not applicable (they'll see a generic "no applicable upgrade found" for all pin related failures)

We may need more work to support messages like:

Package upgrade skipped due to this blocking pin.. Use --force to override
Package upgrade skipped due to this pinning pin.. Upgrade specifically or use --force
Package upgrade skipped due to gated pin.. Use --force to override

florelis replied:
Removed the error code and replaced it by a generic error for when a package is pinned (not specific to blocking)

florelis commented on src/AppInstallerCLICore/Workflows/PinFlow.cpp:
```diff
@@ -57,17 +57,26 @@ namespace AppInstaller::CLI::Workflow
-            availableVersion->GetProperty(PackageVersionProperty::SourceIdentifier).get() });
-        if (pin)
+
+        auto packageVersionKeys = package->GetAvailableVersionKeys(PinBehavior::IgnorePins);
```
This function doesn't return a reference. It returns a new vector, so we take ownership of it.

florelis commented on src/AppInstallerCommonCore/Public/AppInstallerVersions.h:
```diff
@@ -172,16 +172,23 @@ namespace AppInstaller::Utility
     // A range of versions indicated by a version and optionally a wildcard at the end.
     struct GatedVersion
```
I don't know. I don't want to have to reason around, for example, order relative to other versions or gated versions, which I'd probably have to if it is a Version. And I don't think we need to.

yao-msft replied:
Ah I see. Got it.

florelis commented:
I moved the logic to filter out pinned versions out of CompositePackage's GetAvailableVersionKeys() and GetAvailableVersion(). Instead, now the version keys returned from GetAvailableVersionKeys() include the pinned state for that version. Versions from different sources may have different pin types, and versions that match a gating pin are not marked as pinned. Callers can the use that info to decide what to do.

I added another function like GetAvailableVersion() that also returns the pinned state.

I kept the PinBehavior argument for GetLatestAvailableVersion() because the logic there is more complicated and so it is harder to have each caller do it.

I'm still missing using the --force argument, and the messages that I'm exposing could probably be more specific.

florelis commented on src/AppInstallerRepositoryCore/CompositeSource.cpp:
```diff
@@ -467,30 +603,53 @@ namespace AppInstaller::Repository
+                            availablePackage.Pin = std::move(pin.value());
+                        }
+                    }
+                    else if (pinningIndex.GetPin(pinKey))
```
This is only reachable from `SearchInstalled()`. Added a comment about it.

florelis commented on src/AppInstallerRepositoryCore/CompositeSource.cpp:
```diff
@@ -430,34 +539,61 @@ namespace AppInstaller::Repository
             }
 
-            void SetAvailablePackage(std::shared_ptr<IPackage> availablePackage)
+            void AddAvailablePackage(std::shared_ptr<IPackage> availablePackage)
```
I made using multiple available packages dependent on the pinning experimental feature. A pin may make us need multiple sources in almost all commands, for example if a gated version hides all the available versions from one source but not from other sources.

I also added a "default" available source that is used for things like TrySetOverrideInstalledVersion, and that is used first as the source of truth in GetProperty.

florelis commented on src/AppInstallerRepositoryCore/CompositeSource.cpp:
```diff
@@ -320,11 +333,88 @@ namespace AppInstaller::Repository
+                : AvailablePackage(availablePackage), Pin(pin)
+            {
+                auto lastAvailable = AvailablePackage->GetLatestAvailableVersion(PinBehavior::IgnorePins);
+                if (lastAvailable)
```
We actually have a test to check that a null available version is handled properly. I'm not sure when that would happen, but I trust there was a reason for it.

florelis commented:
/azp run

yao-msft commented on src/AppInstallerRepositoryCore/CompositeSource.cpp:
```diff
@@ -320,11 +369,113 @@ namespace AppInstaller::Repository
+                }
+            }
+
+            std::string SourceId;
```
Should the 3 member variables be private? and named m_*?

florelis replied:
Do you feel strongly about this?
It was easier for me to keep them public as I use those fields in the composite package, and the Pin one I also set from there.

yao-msft replied:
If it's not a lot of code changes, I think it'd be better to be private. Because this struct is more of a class and less of a data structure now.

But if it's too many changes, I'm good with leaving it as is. Your call.

yao-msft commented on src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:
```diff
@@ -719,17 +722,30 @@ namespace AppInstaller::CLI::Workflow
         auto &source = context.Get<Execution::Data::Source>();
         bool shouldShowSource = source.IsComposite() && source.GetAvailableSources().size() > 1;
 
+        PinBehavior pinBehavior;
```
nit: initialize value

yao-msft commented on src/AppInstallerCLICore/Workflows/WorkflowBase.cpp:
```diff
@@ -875,7 +906,71 @@ namespace AppInstaller::CLI::Workflow
+            // TODO: The logic here will probably have to get more difficult once we support channels
+            if (Utility::IsEmptyOrWhitespace(m_version) && Utility::IsEmptyOrWhitespace(m_channel))
+            {
+                PinBehavior pinBehavior;
```
nit: initialize value

florelis commented on src/AppInstallerCLICore/Workflows/DependencyNodeProcessor.cpp:
```diff
@@ -41,7 +41,18 @@ namespace AppInstaller::CLI::Workflow
         m_nodePackageInstalledVersion = package->GetInstalledVersion();
-        m_nodePackageLatestVersion = package->GetLatestAvailableVersion();
+
+        PinBehavior pinBehavior;
```
The compiler doesn't complain because it can tell that it is set in both branches of the conditional. In my mind it is clearer this way. If I initialize it, in my mind it opens the questions of "when does it stay with this value?", "why is this the default?", neither of which really matters.

I just checked and if it can stay uninitialized, the compiler errors out, so there's no risk of using it uninitialized even if I don't do it here.

florelis commented on src/AppInstallerRepositoryCore/Public/winget/RepositorySearch.h:
```diff
@@ -232,8 +223,9 @@ namespace AppInstaller::Repository
 
-        PackageVersionKey(Utility::NormalizedString sourceId, Utility::NormalizedString version, Utility::NormalizedString channel) :
-            SourceId(std::move(sourceId)), Version(std::move(version)), Channel(std::move(channel)) {}
+        // TODO #219 check all uses
```
Oh, I got the issue numbers confused and then I forgot to remove this todo. It was about checking all the uses of the constructor to make sure that the default value for pin type/state was appropriate. I just did that, so I'll remove it.

yao-msft commented on src/AppInstallerCLICore/Argument.cpp:
```diff
@@ -149,6 +149,8 @@ namespace AppInstaller::CLI
+        case Execution::Args::Type::IncludePinned:
+            return { type, "include-pinned"_liv, "pinned"_liv, ArgTypeCategory::CopyFlagToSubContext };
         case Execution::Args::Type::UninstallPrevious:
             return { type, "uninstall-previous"_liv, ArgTypeCategory::InstallerBehavior };
```
>uninstall-previous [](http://example.com/codeflow?start=28&length=18)

I guess this should be copied to subcontext as well since we support multiple installs using subcontexts?

yao-msft reviewed (approved):
:shipit:

felipecrs commented:
That's great news! I can't wait to test it :)

denelon commented:
We're checking to see if any other PRs are super close to done, and we'll get a preview build out quickly™️.
